GVRf/Framework/framework/src/main/java/org/gearvrf/io/GVRGazeCursorController.java

```diff
@@ -172,5 +167,64 @@ void close() {
             context.unregisterDrawFrameListener(this);
         }
         referenceCount = 0;
+        if(threadStarted) {
+            thread.quitSafely();
+        }
+    }
+
+    private class EventHandlerThread extends HandlerThread {
+        private static final String THREAD_NAME = "GVRGazeEventHandlerThread";
+        public static final int SET_POSITION = 0;
+        public static final int SET_KEY_EVENT = 1;
+        public static final int SET_KEY_DOWN = 2;
+        public static final int SET_KEY_UP = 3;
+        private final KeyEvent BUTTON_GAZE_DOWN = new KeyEvent(
+                KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_BUTTON_1);
+        private final KeyEvent BUTTON_GAZE_UP = new KeyEvent(KeyEvent.ACTION_UP,
+                KeyEvent.KEYCODE_BUTTON_1);
+        private Handler gazeEventHandler;
+        public EventHandlerThread() {
+            super(THREAD_NAME);
+        }
+
+        public void prepareHandler() {
+            gazeEventHandler = new Handler(getLooper(), new Callback() {
+                @Override
+                public boolean handleMessage(Message msg) {
+                    switch (msg.what) {
+                        case SET_POSITION:
+                            float[] position = (float[])msg.obj;
+                            GVRGazeCursorController.super.setPosition(position[0],position[1],
+                                    position[2]);
+                            break;
+                        case SET_KEY_DOWN:
+                            buttonDownSent = true;
+                            setKeyEvent(BUTTON_GAZE_DOWN);
+                            break;
+                        case SET_KEY_UP:
+                            setKeyEvent(BUTTON_GAZE_UP);
+                            break;
+                        case SET_KEY_EVENT:
+                            KeyEvent keyEvent = (KeyEvent)msg.obj;
+                            setKeyEvent(keyEvent);
+                            break;
+                    }
+                    return false;
+                }
+            });
+        }
+
+        public void setPosition(float x, float y, float z) {
+            Message msg = Message.obtain(gazeEventHandler,SET_POSITION,new float[]{x,y,z});
```

Comment:
You can reuse a Vector3 to prevent a float array being created every time setPostion is called.